### PR TITLE
test/stress: sustained-rate (Poisson) phase, monotonic latency accounting, client connection calibration

### PR DESCRIPTION
**Pull request by aditya-shantanu**
#### What this PR does / why we need it:

Adds three harness improvements to the stress tool, all found while benchmarking SandboxClaim adoption against warm pools with a 300-claim workload. No controller behavior changes; test harness only.

**1. `claims-warm-sustained` phase: open-loop, sustained-rate SandboxClaim arrivals (`sustained.go`).**
The existing phases either launch cold sandboxes or fire everything at once, which is an adversarial burst shape. This phase models steady-state arrivals instead: claims arrive as a Poisson process at `--sustained-rate` claims/s for `--sustained-seconds`, each is deleted `--claim-dwell` after Ready, and the warm pool controller refills continuously, so adoption, refill, and teardown churn compound like they would in production. Arrivals are open-loop (paced by absolute schedule offsets, never gated on earlier claims completing), so the phase measures a target rate rather than degrading into a closed loop. The headline output is rolling per-10s-window create->Ready percentiles bucketed by arrival time (`sustainedWindows` in summary.json, also printed live), which makes "does latency hold over time?" directly visible. `--sustained-namespaces=N` optionally spreads pools and claims round-robin over N pre-created namespaces (namespaces are a natural shard key for controller sharding experiments), and `--sustained-pool-headroom` sizes each pool as ceil(rate/namespaces x headroom-seconds) so pool sizing covers the controller's worst-case refill latency. Unit tests verify the Poisson schedule is strictly increasing, realizes the target rate, and is genuinely exponential (CV ~= 1), plus pool sizing and windowed-latency math (including that degradation over time produces increasing window percentiles).

**2. Monotonic, client-side latency accounting for claim records (`tracker.go`, `stats.go`).**
While validating the benchmark we found two ways to silently corrupt latency numbers, and this PR bakes the defenses into the harness: (a) durations must never be derived from server-side `metadata.creationTimestamp` / condition `lastTransitionTime`, which are truncated to whole seconds — in a 300-claim warm-adoption benchmark, timestamp-truncation math inflated reported queue-wait about 30x (reported p50 1025ms vs 33ms measured client-side; see #781 for the same class of bug); and (b) per-record timing must come from unconditional client-side observation (watch-event receipt, `time.Now()`), never from sampled/rate-limited log lines, which drop the slow cohort and bias every derived quantile. The claim tracker therefore stamps all milestones with client monotonic-clock timestamps at watch-event receipt, records the server-side timestamps separately as a cross-check only, and the new `computeTimeToAllReady`/`computeWindowedLatencies` aggregations (with tests) operate purely on the client-observed milestones.

**3. `--client-connections=N`: shard the harness's mutating requests over N HTTP/2 connections (`clientconns.go`).**
client-go multiplexes everything built from one `rest.Config` onto a single HTTP/2 connection, and the apiserver caps ~100 concurrent streams per connection. Historically the stress tool put all traffic (creates, deletes, watches, scrapes) on that one connection, so a create burst wider than ~100 queued inside the harness waiting for a stream: in the same 300-claim benchmark this inflated reported create-ack latency roughly 2-3x with client-side queueing that says nothing about the apiserver or the controller. With N>1, mutating requests round-robin over N dedicated connections (per-shard `Dial` functions, so client-go's transport cache builds distinct transports, using only supported client-go API) while the watches keep their own untouched connection that create bursts cannot congest. The default N=1 preserves today's single-connection behavior exactly, and the harness warns when `--client-connections < ceil(create-concurrency/100)` so runs that measure the congested shape do so knowingly. Tests assert exactly N TCP dials, HTTP/2 on every shard, deterministic round-robin distribution, and that the watch client's connection stays distinct from every create shard.

Also includes a minimal `generate_report.py` fix required by the new phase: the watch-stream scan now pins the envelope schema and keeps `object` as raw JSON instead of `read_json_auto`, because sandboxclaims watch events (whose `status.sandbox` key is absent from the sampled pod/sandbox shapes) made DuckDB's inferred-STRUCT scan fail with `unknown key` and killed the whole report.

**Why there is no latency A/B table on this PR:** behavior-affecting PRs in this series each carry a BASE-vs-change A/B benchmark run; this PR changes only the test harness, so a controller A/B table would measure the same controller twice. The quantitative evidence here is about measurement honesty, with numbers above: ~30x inflation of reported queue-wait from 1s-truncated server timestamps (reported p50 1025ms vs 33ms measured client-side) and ~2-3x inflation of reported create-ack latency from single-connection client-side stream queueing — both measured in a 300-claim warm-adoption benchmark and cross-checked against the client watch stream.

#### Which issue(s) this PR is related to:

Ref #403, Ref #624 (benchmark/scale-testing umbrellas: this extends the stress harness with a steady-state workload shape and removes two measurement artifacts that misattributed client-side effects to the control plane). Ref #781 (the 1s-truncated-timestamp measurement hazard; this PR keeps the harness immune to it on the claim path). Happy to rebase this onto PR #1148's harness split whenever it lands.

#### Release Note

```release-note
NONE
```

## Summary by CodeRabbit

* **New Features**
  * Added a new `claims-warm-sustained` sustained-rate warm-pool stress phase (now in the default phase list).
  * Introduced configurable sustained workload parameters (arrival rate/duration, claim dwell, multi-namespace sharding, pool headroom, lifecycle budget).
  * Added rolling windowed analytics and a dedicated report section with arrival vs ready counts and create→ack/create→Ready latency percentiles.
  * Extended cluster capacity checks for sustained steady-state warm pools and in-flight claims; added `--claim-dwell >= 0` validation.

* **Tests**
  * Added unit tests covering Poisson scheduling, sustained scaling/capacity calculations, and windowed latency analytics (including edge cases).

### CODE REVIEW

**Comment by aditya-shantanu** [comment]
Rebased onto 0e5bdbc to resolve conflicts with the recent stress-harness changes (#1230/#1235/#1236). No functional changes to this PR's three features; re-expressed over the new claims-warm phase and declared-schema report code — the clientconns and generate_report.py hunks are now fully covered upstream, so this PR shrinks to the sustained phase and the windowed-latency accounting.

File: test/stress/sustained.go [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+	phase := PhaseClaimsWarmSustained
+
+	// Delete claims that never made it to their own delete (failed, timed
+	// out, or the phase aborted mid-dwell).
+	var leftovers []types.NamespacedName
+	for _, rec := range s.tracker.Records() {
+		if rec.PhaseNumber == number && !rec.CreateReturned.IsZero() && rec.SandboxDeleted.IsZero() && rec.DeleteCalled.IsZero() {
+			leftovers = append(leftovers, types.NamespacedName{Name: rec.Name, Namespace: rec.Namespace})
+		}
+	}
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply test/stress/sustained.go:401]: Fixed in 3908f6e — cleanupSustained now deletes every created claim that was never observed deleted, including ones whose delete was attempted but failed (or whose DELETED event the watch missed); re-deleting an already-gone claim is just a cheap NotFound.

File: test/stress/main.go [hunk 2]
```diff
@@ -594,10 +627,26 @@ func buildPhaseRuns(test *stressTest) ([]phaseRun, error) {
...
+			}
+			if test.cfg.SustainedNamespaces < 1 {
+				return nil, fmt.Errorf("phase %q requires --sustained-namespaces >= 1", raw)
+			}
+			if test.cfg.SustainedPoolHeadroom <= 0 {
+				return nil, fmt.Errorf("phase %q requires --sustained-pool-headroom > 0", raw)
+			}
+			runs = append(runs, phaseRun{number, PhaseClaimsWarmSustained, func(ctx context.Context) error {
+				return test.runClaimsWarmSustainedPhase(ctx, number)
+			}})
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply test/stress/main.go:660]: Fixed in 3908f6e — the claims-warm-sustained branch now rejects NaN/±Inf for --sustained-rate and --sustained-seconds (NaN even passes a plain <= 0 check), keeping the existing positive-value requirement and error style.

File: test/stress/main.go [hunk 3]
```diff
@@ -666,6 +720,17 @@ func checkClusterCapacity(cfg Config, info *ClusterInfo) error {
...
+		// Steady state: the pools stay full (refill replaces adopted
+		// sandboxes) while adopted sandboxes live for roughly the ready
+		// latency + dwell + deletion pipeline before their pods go away;
+		// budget ~5s of pipeline on top of the dwell.
+		poolTotal := sustainedPoolReplicasPerNamespace(cfg) * cfg.SustainedNamespaces
+		inFlight := int(math.Ceil(cfg.SustainedRate * (cfg.ClaimDwell.Seconds() + 5)))
+		if n := poolTotal + inFlight; n > extra {
+			extra = n
+		}
+	}
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply test/stress/main.go:749]: Fixed in 3908f6e — the hard-coded 5s allowance is now an explicit --sustained-lifecycle-budget duration flag (default 5s, validated > 0) used by checkClusterCapacity, so a slower Ready/delete path can be budgeted for instead of silently underestimating pods; added TestCheckClusterCapacitySustained covering the pass/fail boundary and the slower-lifecycle case. Kept it an explicit budget rather than deriving from --per-sandbox-timeout (5m default): that is a worst-case abort bound, and sizing capacity to it would demand ~60x the steady-state pods and spuriously fail every run.

**Comment by aditya-shantanu** [comment]
Post-filing stress validation of this harness (run as one batch validating several open PRs, each on its own leg of a 12-node tuned kops cluster; phases `fill,throughput-mif50,throughput-mif100,claims-warm-sustained` at 45 claims/s x 60s, `--claim-dwell=2s`, 4 namespaces):

- This PR's own sustained (Poisson) phase drove every leg's load; on the stock-controller leg it delivered 2,785 open-loop arrivals, **2,785/2,785 Ready, 0 failed**, with the rolling 10s windows stable (p50 114-240ms, no monotonic drift) — the windowed accounting and open-loop pacing behaved as designed under real cluster conditions.
- fill (120) and both closed-loop throughput phases (787 + 1,455 sandboxes) also completed with zero failures.
- Zero cold starts; controller log noise was limited to the known upstream sandbox-status update conflicts, present at the same rate on the stock leg.

Artifacts (per-leg RESULTS, summary.json, metrics.jsonl.gz) retained; happy to attach specifics on request.

File: test/stress/sustained.go [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+}
+
+// cleanupSustained removes everything the phase created: leftover claims
+// (most were already deleted after their dwell), the pools and templates,
+// and — for multi-namespace runs — the extra namespaces themselves. Like the
+// probe/throughput deletes, failures are logged rather than returned so
+// cleanup problems do not mask the phase's measurement, and the end-of-run
+// namespace deletion is the backstop for the main namespace.
+func (s *stressTest) cleanupSustained(ctx context.Context, number PhaseNumber, namespaces []sustainedNamespaceClients, poolName, templateName string) {
+	if ctx.Err() != nil || len(namespaces) == 0 {
```
Comment (janetkuo) [inline test/stress/sustained.go]: If `ctx.Err() != nil` (e.g. because the phase timed out or was cancelled), this returns early and skips cleaning up the extra namespaces. Since these namespaces are created outside the main test namespace, they will leak. We should ensure they are deleted even if the context is cancelled.

**Review by janetkuo — COMMENTED** [review]
Potential namespace leak that could eat up cluster resource. lgtm otherwise.

File: test/stress/sustained.go (hunk 4 above)
Comment (aditya-shantanu) [reply test/stress/sustained.go]: Fixed in 2eadd51 — cleanup now runs on a context detached from the cancelled phase context (`context.WithTimeout(context.WithoutCancel(ctx), 2*time.Minute)`, the same bounded pattern the end-of-run namespace backstop uses), so the extra `-sN` namespaces are deleted even when the phase times out or is cancelled. For single-namespace runs everything lives in the main test namespace, so that case still defers to the end-of-run backstop.

I also checked the sibling early returns with the same shape (`cleanupClaimsWarm` and `deleteSandbox` in phases.go): those only touch resources inside the main test namespace, which the backstop covers, so they are left as-is.

Good catch, thank you @janetkuo — these namespaces had no backstop at all.

**Review by janetkuo — APPROVED** [review]
/lgtm
